docs_src/design/adr/0026-Common Configuration.md

```diff
@@ -0,0 +1,401 @@
+
+During bootstrapping, each service will either load the common configuration from the Configuration Provider or via URI to some endpoint that provides the common configuration. This could be the new service's REST endpoint. Each service will have additional private configuration, which may overwrite and/or extend the common configuration.  
+
+An additional common configuration setting must be present to indicate all other common settings have been pushed to the Configuration Provider. This setting is stored last and the services must wait for this setting to be present prior to pulling the common settings.
+
+### Configuration Bootstrapping
+
+The following flow chart demonstrates the bootstrapping of each services' configuration with this new common configuration capability.
+
+![](common-config-images/EdgeX 3.x Configuration bootstrapping flowchart.png)
+
+### Secret Store Configuration
+
+As part of this design, the Secret Store configuration is being removed from the service configuration (common and private). This is so the Secret Provider can be instantiated prior to processing the service's configuration which may require the Secret Provider. The Secret Store configuration will now be a combination of default values and environment variable overrides. These environment variables will be the same as the ones that are currently used to override the configuration.
+
+### Specifying the Common Configuration location
+
+If the `-cp/--configProvider` command line option is used, the service will default to pulling the common configuration from a standard path in the Configuration Provider. i.e. `edgex/3.0/common/` The `-cp/--configProvider` option assumes the usage of the core-common-config service and cannot be used with the `-cc/--commonConfig` option.
+
+The new `-cc/--commonConfig` command line option will be added for all services. This option will take the URI that specifies where the common configuration is pulled when not using the Configuration Provider. If the `-cc/--commonConfig` option is not specified, the URI will default to the new `core-common-config` service's REST API. Authentication will be limited to `basic-auth`. In addition, a new environment override variable `EDGEX_COMMON_CONFIG` will be added which allows overriding this new command line option.
+
+#### Options for Providing the Common Configuration
+
+1. From the configuration provider, using the `-cp/--configProvider` command line option
+2. From the new `core-common-config` service, either using `-cc/--commonConfig` command line option, or if `-cp/--configProvider` and `-cc/--commonConfig` are not specified, the URI will default to the new `core-common-config` service`s REST API.
+3. For the simplest setup, without the `-cp/--configProvider` and `core-common-config` service, the `-cc/--configProvider` option may be used to provide the URI for a text file. This simple approach has no authentication and no update mechanism.

```

Comment:
```suggestion
1. From the Configuration Provider, using the `-cp/--configProvider` command line option
2. When Configuration Provider is not used, given the URI in `-cc/--commonConfig` command line option:
    1. From an HTTP endpoint that returns the common configuration file in response
    2. From a local file that contains the common configuration
```

Reply:
So do we add a #3 and #4 like:
3. When Configuration Provider is not used, a new environment override variable `EDGEX_COMMON_CONFIG` can be used to override the `-cp/--configProvider` option and provide the URI as described in 2. This variable can also provide the URI for all services if the `-cp/--configProvider` option isn't provided.
4. If common configuration isn't provided by one of the 3 means above, then the services cannot start.

Reply:
@farshidtz This will make `-cc` required when `-cp` is not used. 

I think it would be better to have a default for ease of use. There are two paths we could take for the default case (when `-cc` and `-cp` are not provided):

1. Use the new service to provide the default. (what is currently in the ADR)
2. Use the private configuration only. This would allow for legacy support by default. However, all our reference services will be updated to use common configuration and the private configurations will not have all the necessary sections. So when users run our reference services from command line, they will get errors because the configurations are incomplete. This would lead to more questions in Slack and GitHub. 

@TomBrennan-Eaton Yes, we can add 3 and 4 if we decide to go without a default.

@cloudxxx8 @iain-anderson Thoughts?

Reply:
IMO, having the the new `core-common-config` service be the source for the common config for Consul and for Non-Consul in our reference deployments is elegant. All service's running w/o Consul will default to this new service for the common config. Users the use `-cc` when they have a different source they want to use.

The service is there, it owns the common config, why not have it be the default source when Consul isn't used?

Reply:
@lenny-intel Farshid pointed out that if the REST API is just serving up a file (in effect) that this could be done simply enough by just having something (eg, nginx) host the file (locally or remotely). Then you don't need it to support a REST API for common config.
And taking away the REST API reduces it to just a simple "start up, run once, and exit" task, so that it would not be necessary for it to be an EdgeX Service, with all that that entails.
OK, agree to "elegant", but (I think I can say for Farshid) and I think we should minimize this common config provider to less than a service.
That left us with the "what should the *default* URI be?" problem, and Iain pointed out that there was no good answer to that. 
I suggested that in simple systems where you don't have a config provider (and probably don't want to have a common-config service either, to minimize the footprint) that having a file directly accessible or hosted seems like a suitably good solution.
Finally, this is what I really like about the EDGEX_COMMON_CONFIG environment variable - it's a "good-enough-answer" to the "what should the *default* URI be?" question, for simple systems - easy to build into a simple system.

Reply:
@ejlee3 When CP is used, the new service should not run as a service by default so that we don't waste system resources. In this case, if CP is NOT used, the default URI will NOT help with the startup. A user going out of the "default" path would have to remove CP dependency and provide the URI to common config (file or served over the network).

> So when users run our reference services from command line, they will get errors because the configurations are incomplete.

Not if they are using CP, which is the default.

@lenny-intel @ejlee3 As discussed in the call, having it as a service has design flaws:
1. Introducing a new built-in service probably needs additional considerations (footprints) and TSC approval. @JamesKButcher @jpwhitemn what do you think?
2. The given design for a service it isn't secure to the same level as other EdgeX services. First, it is HTTP by default and not routed through the API Gateway. It will be at the risk of [man in the middle attack](https://en.wikipedia.org/wiki/Man-in-the-middle_attack)). Moreover, the optional access control won't be consistent with any other services in EdgeX, because it is Basic Auth (which should be used only with TLS!) vs whatever we end up having for other services. @bnevis-i are you okay with this approach and approve it from a security standpoint, considering [this UCR](https://github.com/edgexfoundry/edgex-docs/pull/909/files)?
3. The service is a stripped down Configuration Provider. IMO, the alternative to no-CP should not be a stripped-down CP, but instead to have no EdgeX service serving configurations over the network.
 
I don't understand the need to make EdgeX configuration architecture more complicated than it already is. What is needed to satisfy the UCR is to allow supply of common config from CP or from a URI. I don't see any approved requirement to have a common config provider service. See https://docs.edgexfoundry.org/2.4/design/ucr/Common%20Configuration/

Reply:
@farshidtz according to this ADR, CP still needs the common-config service to push to common part of the configuration, or we need another mechanism to initialize the common configs.

The configuration of microservice architecture is complicated, so we need this mechanism to make it easier to use although it's more complicated for the developers like us.
Users usually want to change the config for the whole system, but they aren't quite sure which service config should be change.  Common config here is helpful from users' point of view.

Reply:
I am fine with not having the -cc default value.

Reply:
Addressing each point individually:

> The given design for a service isn't secure to the same level as other EdgeX services. First, it is HTTP by default and not routed through the API Gateway; it will be at the risk of [man in the middle attack](https://en.wikipedia.org/wiki/Man-in-the-middle_attack). 

All EdgeX services today are subject to MITM attacks and the mitigation is still the same if that threat is in scope--TLS at the application level, encrypted tunnels, or service mesh/fabric.  I don't see how a proposed common configuration service is any different.

> Moreover, the optional access control won't be consistent with any other services in EdgeX, because it is Basic Auth (which should be used only with TLS!) vs whatever we end up having for other services. 

Yes, this bugs me.  Per the [TLS ADR](https://docs.edgexfoundry.org/2.3/design/adr/security/0015-in-cluster-tls/) we aren't concerned about Basic Auth (or token auth for that matter) occurring over non-TLS channels if communication is on the same host.  If you are root, you can capture packets and retrieve credentials, or steal TLS keys and spoof the receiving service.

Moreover, despite the common configuration, in order to perform authentication, the credential is coming from the EdgeX secret store.  **Thus, the secret store configuration must still come from disk and not a common configuration provider service.**  There must still be some local configuration available to each EdgeX service that tells it parameters by which it trusts and authenticates to the EdgeX secret store.

@bnevis-i are you okay with this approach and approve it from a security standpoint, considering [this UCR](https://github.com/edgexfoundry/edgex-docs/pull/909/files)?

I would rather the ADR state at the authentication mechanism be identical to the one chosen for other EdgeX peer microservices, instead of going off on its own.  I presume Basic Auth was chosen because it could be implemented easily, and because common configuration would be needed to successfully initialize go-mod-bootstrap, and Basic Auth was an easy way to avoid a chicken-and-egg problem.

Reply:
> Moreover, the optional access control won't be consistent with any other services in EdgeX, because it is Basic Auth (which should be used only with TLS!) vs whatever we end up having for other services.

No where did we state the this reference endpoint would require basic auth. What we did state is that loading of the file from URI would support use of basic auth. This doesn't mean it would require it.

> A good point mentioned earlier is that if we are simply serving a file with minimal authentication, Nginx, mini_httpd, or the like will probably work. python's http.server doesn't support auth. The more complex the auth mechanism, the fewer the choices.

This new service already owns the file for the purposes of pushing it into Consul. I don't see how adding Nginx or something else to serve the file it is simpler.

Note that in secure mode this new service will need a Vault token and a Consul Access token in order to push the configuration into Consul. Also all services will need ACL access to the common area in Consul. @ejlee3, this should be added to the ADR.

Reply:
> This new service already owns the file for the purposes of pushing it into Consul. I don't see how adding Nginx or something else to serve the file it is simpler.
> 
> Note that in secure mode this new service will need a Vault token and a Consul Access token in order to push the configuration into Consul. Also all services will need ACL access to the common area in Consul. @ejlee3, this should be added to the ADR.

@lenny-intel You are making this argument because this new service is either serving the config directly, or pushing it into Consul, or both, and that since you still need to push to Consul, it has to be a full service.  Correct?

Reply:
> 2\. The given design for a service isn't secure to the same level as other EdgeX services. First, it is HTTP by default and not routed through the API Gateway;

Since this new service is present only for the local EdgeX services (reference deployment), there is no need for it to be accessible via the API gateway.

Reply:
> Since this new service is present only for the local EdgeX services (reference deployment), there is no need for it to be accessible via the API gateway.

@lenny-intel This is not true. The services can run on different machines. Running a device service **securely** on a different machine has been a long requested feature and this is just a step backward in making that possible.

Reply:
> This new service already owns the file for the purposes of pushing it into Consul. I don't see how adding Nginx or something else to serve the file it is simpler.

@lenny-intel the new component doesn't need to be a long running service. All is needed per requirements is to simply seed the CP and exit. When CP isn't used, the commons can be loaded from a file as for having the simplest approach. If you make it run forever, you would be wasting memory and CPU cycles for no reason.

Reply:
> Note that in secure mode this new service will need a Vault token and a Consul Access token in order to push the configuration into Consul. @ejlee3, this should be added to the ADR.

> You are making this argument because this new service is either serving the config directly, or pushing it into Consul, or both, and that since you still need to push to Consul, it has to be a full service. Correct?

> correct it will need most of the "full service" capability to handle pushing to Consul. So why not also have it be a default source for the file for when Consul isn't used. 

@lenny-intel @bnevis-i Sorry, but again, why does it need to maintain a valid Vault token when its role is to bootstrap the file once. Is it needed to watch and re-submit the file automatically once changed? (where is the requirement?) If it is to submit once on runtime, I don't see the need for it to be a service to keep renewing the Vault token. If the config file has been modified by the user, that same user can run the security setup once to get a new and valid token and then run the application to seed and exit.

Reply:
> If you make it run forever, you would be wasting memory and CPU cycles for no reason.

It causes a problem for people running EdgeX in a monolith Kubernetes pod.  If one container terminates, the whole POD fails.

Reply:
> Sorry, but again, why does it need to maintain a valid Vault token when its role is to bootstrap the file once.

@farshidtz , In secure mode Consul is locked down so the service will need a Vault token in order to request a Consul Access token. After it pushes the common config it no longer needs the tokens.

Reply:
> Not if they are using CP, which is the default.

@farshidtz , this is not the default when running from command-line in a terminal. One has to explicitly use `-cp` option.

Reply:
> @lenny-intel So that sounds like a one-shot service which waits until the common config has been submitted does the job. It will have to maintain the Vault token in the meantime as per @bnevis-i's concern.

It will need to maintain the token if:
- It continues to run a server that serves configuration for those components that don't fetch it from consul.
- It wishes to participate in standard EdgeX service authentication 

So if it continues running with a cached HTTP Basic auth credential, or terminates and just dead hangs the container (this is the one-shot scenario), no token maintenance is required.

Reply:
> The `-cc` introduced in this ADR takes the above mentioned URI. The authentication is limited to basic auth because the credentials can be passed in a URI in form of `http://user:pass@myserver.com/common.json`. There is no requirement to have the server of that file participate in EdgeX service auth. I don't know would on implement that, since we only have a URI without any other information about the target endpoint (Are we gonna send the secret to whatever server that serves the file, even if it is github.io?)

It might be beneficial to the spec, that if HTTPS is used, to specify how that trust is established.  (My assumption would be to keep it simple, that it must be trusted by the system, otherwise another command line argument to specify the certfile or certpath.).   Also, I want to note that the  `http://user:pass@myserver.com/common.json` form, if specified on the command line, is insecure (it will show up in `ps` output).  `curl` for example, has a `--netrc` option to avoid that.